content/en/BestPractices/defend-against-ssrf.md

```diff
@@ -0,0 +1,257 @@
+
+    return urllib.request.urlopen(url).read() 
+
+```
+
+
+
+In the above code, the application accepts the user-supplied input through the *url* parameter and passes it to the *urllib.request.urlopen(url).read()* function. SSRF attacks are possible since the malicious URLs supplied by the users are accepted and processed without filtration. 
+
+
+
+Example: https://www.target.com/test/?url=http://127.0.0.1 will allow an attacker to access the `127.0.0.1 (localhost)` by passing it to the vulnerable code. 
+
+
+
+Below is the secure alternative of this code where we use a `Whitelist` that will be matched against the user-supplied URL. The application will only process the request if the domains in the allowlist match the user-supplied URL. 
+
+
+
+However, in this case, the `127.0.0.1` from the above example does not match `domain1.com` and `domain2.com`. Hence it would not execute the request. 
+
+
+
+
+
+```
+
+from flask import request
+
+import urllib
+
+
+
+DOMAINS_WHITELIST = ['domain1.com', 'domain2.com']
+
+
+
+@app.route('/proxy')
+
+def proxy():
+
+    url = request.args["url"]
+
+    if urllib.parse.urlparse(url).hostname in DOMAINS_WHITELIST:
+
+        return urllib.request.urlopen(url).read()
+
+```
+
+---
+
+### Scenario - 2: SSRF through Use of a Known Vulnerable Function
+
+
+
+Often developers use known frameworks and functions to implement specific functionalities. However, many functions and methods are not secure by default and could unintentionally inhibit potentially sensitive vulnerabilities like SSRF. 
+
+
+
+In the below code, the `URI.resolve` method is used. If an absolute URL is supplied, it will return its parameter. An attacker can abuse this behaviour to perform a successful SSRF attack. 
+
+
+
+```
+
+URI BASE URI = URI.create("https://repo.maven.apache.org/maven2/") ;
+
+@GetMapping("/pkg/{group}/fartifact}/{version}")
+
+public ResponseEntity Resource> getPackagel
+
+        @Pathvariable String group,
+
+        @PathVariable String artifact,
+
+        @PathVariable String version) {
+
+    Path pkgPath = Paths.get (String.join("/", group.split("\\.")))
+
+            .resolve(artifact)
+
+            . resolve (version)
+
+            . resolve (String. format ("%S-s.jar", artifact, version))
+
+            . normalize() ;
+
+    URI pkgUrl = BASE _URI.resolve (pkgPath.toString());
+
+    return sendResponseStream(this.okHttpWrapper.get(pkgUrl.toString()));
+
+}
+
+```
+
+It is recommended to analyze the methods used to handle user-supplied input for any known security concerns and avoid using them. In case there is no alternative, writing a wrapper around it is recommended for further filtration and preventing an attack from execution. 
+
+---
+
+### Scenario - 3: Extracting Cloud Metadata 
+
+
+
+Suppose the application is powered by Cloud services such as AWS, Digital Ocean, Google Cloud, etc... In that case, An attacker may attempt to leverage an SSRF issue to steal the cloud metadata credentials. 
```

Comment:
`For applications that are powered by Cloud services, such as..., an attacker could attempt to leverage SSRF`